### PR TITLE
STY: Consolidate & add pre-commit checks

### PR DESCRIPTION
**Pull request by mroeschke**
* Adds some easy/useful `pre-commit-hooks` and `pygrep-hooks`
* Consolidates & de-duplicates some redundant pre-commit `pygrep` checks

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -197,16 +193,8 @@ repos:
 
             # builtin filter function
             |(?<!def)[\(\s]filter\(
-
-            # exec
-            |[^a-zA-Z0-9_]exec\(
         types_or: [python, cython, rst]
         exclude: ^doc/source/development/code_style\.rst  # contains examples of patterns to avoid
-    -   id: cython-casting
```
Comment (MarcoGorelli) [inline .pre-commit-config.yaml:205]: is this one covered by anything else?

**Review by MarcoGorelli — APPROVED** [review]
Nice, looks good on green

File: .pre-commit-config.yaml (hunk 1 above)
Comment (mroeschke) [reply .pre-commit-config.yaml:205]: This one is in `id: unwanted-patterns-in-cython` (would be nice if this was in cython-lint!)

**Comment by mroeschke** [comment]
Failures are unrelated here, so merging

**Pull request merged**